### ISSUE TITLE
Interface imports not parsed out

### ISSUE THREAD

**Issue by mightyiam** [eslint/typescript-eslint-parser#179]
**What version of TypeScript are you using?**
See `package.json`.

**What version of `typescript-eslint-parser` are you using?**
See `package.json`.

**What code were you trying to parse?**

See repository:
https://github.com/mightyiam/cyclejs-web-components/pull/13

**What did you expect to happen?**

Interface imports should have been parsed out.

**What happened?**

They seem to have remained, because I got:
https://travis-ci.org/mightyiam/cyclejs-web-components/jobs/207631830#L658, for example.

**Comment by soda0289** [comment]
I don't have a lot of free time to dig into this. Could you post a short code sample that reproduces the failure. The output in travis does not say what rule is failing so it is hard to diagnose the problem. Could you use the eslint from the command line and post the output from there.

If I had to guess it looks like the no-unused-vars or no-undef rule. Both of these rules are known to not currently work due in part to the way eslint and escope traverse the AST. You can check out this issue https://github.com/eslint/typescript-eslint-parser/issues/77 for more information on rules that do not work woth this parser. There are some workaround such as using the typescript compiler options noUnusedLocals and noUnusedParameters. There is also a rule in the typescript-eslint-plugin but it does not cover all cases.

**Comment by QuantumInformation** [comment]
For no-unused-vars I am getting false positives for valid ts code:

`import { IGameUnit } from "../gameUnits/IGameUnit";`

which is used:

![image](https://cloud.githubusercontent.com/assets/216566/24326920/de014dfe-11b1-11e7-99dc-18eb3a266a6f.png)

any ideas?

**Comment by soda0289** [comment]
@QuantumInformation This is a known issue. We are working on making eslint traverse type annotations and decorators to allow for the rule no-unused-vars to work as expected.

Check out this issue for progress:
https://github.com/eslint/eslint/issues/7129

I have created a patch to  get this to work with eslint-scope.
https://github.com/eslint/eslint-scope/pull/26

 We currently use a rule in typescript eslint plugin that can mark decorators as used  but not type annotations. Either way once eslint can traverse typeannotions we can update the eslint plugin to mark them as used or use new escope logic.

**Comment by QuantumInformation** [comment]
thanks guys

**Comment by sompylasar** [comment]
@soda0289 RE: https://github.com/eslint/typescript-eslint-parser/issues/179#issuecomment-289310710

This https://github.com/eslint/eslint/issues/7129 has been merged in https://github.com/eslint/eslint/pull/8365

This https://github.com/eslint/eslint-scope/pull/26 has been closed [in favor of creating plugins/hooks](https://github.com/eslint/eslint-scope/pull/26#issuecomment-291001599).

What's the status of this? `no-undef` is really annoying on `interface`s. Thank you!

**Comment by soda0289** [comment]
@sompylasar The issue of traversing type annotations or decorators will not fix the problem with `no-undef` or `no-unused-vars`. The problem lies in eslint-scope and not eslint or estraverse. My original assumptions about how eslint traversed nodes was incorrect.

For this to be fixed we need eslint-scope to understand typescript nodes that are produced by this parser's AST. The babel-eslint parser currently monkey patches eslint-scope to achieve this by overriding methods in the scopers referencer class. This solution is not ideal and I have suggested allowing the parser to pass its own set of VisitorKeys and eslint-scope node listeners or plugins.

The first stage of passing VisitorKeys to eslint-scope and estraverse will allow us to only traverse nodes that we know are handled correctly by eslint and eslint-scope. This will allow for `no-undef` to not throw false positives. This PR is waiting for TSC approval and review. https://github.com/eslint/eslint/pull/8582

The second stage of fixing eslint-scope is a little more complicated. One solution is to add hooks into visitor functions and allow for the parser to override node listeners. https://github.com/eslint/eslint-scope/issues/30. This is not perfect but will allow us to replace the monkey patch used in babel-eslint and fix issues with `no-undef` and `no-unused-vars`. A better solution would be to allow for multiple node listeners to exists and for plugins to be created for things like decorators, type annotations, and type parameters. This needs more thought and would require a bigger refactor of the eslint-scope code base.

There has been some other ideas of allowing parsers to supply their own scope anaylsis tool but this should all be discussed in the next TSC meeting and hopefully a more concrecet plan will come forward.

**Comment by sompylasar** [comment]
@soda0289 Thank you for the detailed explanation! The unfortunate thing here is this will likely happen so far in the future, even though so many people are willing to help out... and [eslint@4 will have to be adapted by the most popular rule template and every of its dependencies](https://github.com/airbnb/javascript/issues/1421).

**Comment by soda0289** [comment]
Yes it will take time for plugins to update but ESLint v4 brings many fixes and features to this parser besides the possibility of better scope analysis. We could look into creating a monkey patch now for eslint v3 support but it only effects two rules and both of those rules can be enforced by the typescript compiler itself. 

Another thing to keep in mind is that babel is working towards typescript support as well. Meaning it might be possible to use babel-eslint with typescript and eslint v3 depending on how things work out.

**Comment by sompylasar** [comment]
> We could look into creating a monkey patch now for eslint v3 support but it only effects two rules and both of those rules can be enforced by the typescript compiler itself.

That would be great actually. I'm currently trying to introduce TypeScript into JavaScript project while not transforming the whole project to TypeScript, so I need that support from ESLint, not from TypeScript.

> Another thing to keep in mind is that babel is working towards typescript support as well. Meaning it might be possible to use babel-eslint with typescript and eslint v3 depending on how things work out.

And I've succeeded in doing this right now (versus "might be possible") as I see immediate benefits of using TypeScript on the large-scale project I'm working on, so [there is no time to wait](https://github.com/benmosher/eslint-plugin-import/issues/839), we have to [push the boundaries](https://github.com/benmosher/eslint-plugin-import/pull/840).

I'd rather made tweaks to ESLint config ruleset based on the file type it's applied for, but that is not possible because ESLint does not provide the `filePath` to the config, assuming the config should be static and one for all, so the only thing I can do is to alter the parser, but not the rules.

```js
// .eslintrc.js
  "parser": "./webpack/parserForEslint.js",
```
```js
// ./webpack/parserForEslint.js
module.exports = {
  parse: function (text, parserOptions) {
    if (/\.tsx?$/.test(parserOptions.filePath)) {
      return require('typescript-eslint-parser').parse(text, parserOptions);
    }
    return require('babel-eslint').parse(text, parserOptions);
  },
};
```

**Comment by soda0289** [comment]
I'm not convinced we should add a monkey patch just for two rules but if no other solution can be reached we might be forced into it.

If you are looking for a better solution for using different parsers/rules for different files work is being done in eslint to make this happen.
https://github.com/eslint/eslint/pull/8081
https://github.com/eslint/eslint/issues/8543

We should probably start a new issue if you want to discuss this further.

**Comment by sompylasar** [comment]
And we've got [the negative response from airbnb](https://github.com/airbnb/javascript/issues/1421#issuecomment-302968881), they're not going to upgrade to `eslint@4` in observable future. So if we really want to enable JavaScript people moving to TypeScript to use eslint (this repo's purpose), a hacky patch now is better than nothing now and something better in the distant future.

**Comment by sompylasar** [comment]
I ended up cutting `TSModuleDeclaration` nodes out of the AST to prevent `no-undef` from triggering within them:
```
module.exports = {
  parse: function (text, parserOptions) {
    if (/\.tsx?$/.test(parserOptions.filePath)) {
      const ast = require('typescript-eslint-parser').parse(text, parserOptions);

      // WORKAROUND(@sompylasar): Drop `declare module` to prevent its contents from triggering `no-undef`. https://github.com/eslint/typescript-eslint-parser/issues/77#issuecomment-293581001
      ast.body = ast.body.filter(function (node) {
        return (node.type !== 'TSModuleDeclaration');
      });

      return ast;
    }
    return require('babel-eslint').parse(text, parserOptions);
  },
};
```

**Comment by soda0289** [comment]
Closing this in favor of #77.
